credential/src/main/java/io/syndesis/credential/CredentialFlowState.java

```diff
@@ -17,27 +17,91 @@
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
 
-@Value.Immutable
-@JsonDeserialize(builder = CredentialFlowState.Builder.class)
+import org.springframework.social.connect.ConnectionFactory;
+import org.springframework.social.connect.support.OAuth1ConnectionFactory;
+import org.springframework.social.connect.support.OAuth2ConnectionFactory;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+@JsonSubTypes({@JsonSubTypes.Type(value = ImmutableOAuth1CredentialFlowState.class, name = "OAUTH1"),
+    @JsonSubTypes.Type(value = ImmutableOAuth2CredentialFlowState.class, name = "OAUTH2")})
 public interface CredentialFlowState {
 
-    class Builder extends ImmutableCredentialFlowState.Builder {
-        // builder implemented by Immutables, access allowed through this
-        // subclass
+    String CREDENTIAL_PREFIX = "cred-o";
+
+    String OAUTH1_CREDENTIAL_PREFIX = CREDENTIAL_PREFIX + "1-";
+
+    String OAUTH2_CREDENTIAL_PREFIX = CREDENTIAL_PREFIX + "2-";
+
+    interface Builder {
+
+        CredentialFlowState build();
+
+        Builder connectionId(String connectionId);
+
+        Builder key(String key);
+
+        Builder providerId(String providerId);
+
+        Builder returnUrl(URI returnUrl);
+
     }
 
     String getConnectionId();
 
+    String getKey();
+
     String getProviderId();
 
     URI getReturnUrl();
 
-    String getKey();
+    default String persistenceKey() {
+        return statePrefix() + getKey();
+    }
+
+    default String statePrefix() {
+        if (this instanceof OAuth1CredentialFlowState) {
+            return OAUTH1_CREDENTIAL_PREFIX;
+        } else if (this instanceof OAuth2CredentialFlowState) {
+            return OAUTH2_CREDENTIAL_PREFIX;
+        }
+
+        throw new IllegalStateException("Unsupported credential flow state implementation: " + this);
+    }
+
+    default Type type() {
+        if (this instanceof OAuth1CredentialFlowState) {
+            return Type.OAUTH1;
+        } else if (this instanceof OAuth2CredentialFlowState) {
+            return Type.OAUTH2;
+        }
+
+        throw new IllegalStateException("Unsupported credential flow state implementation: " + this);
+    }
 
-    OAuthToken getToken();
+    CredentialFlowState updateFrom(HttpServletRequest request);
+
+    static Builder builderFor(final ConnectionFactory<?> connectionFactory) {
+        if (connectionFactory instanceof OAuth1ConnectionFactory) {
+            return new OAuth1CredentialFlowState.Builder();
+        } else if (connectionFactory instanceof OAuth2ConnectionFactory) {
+            return new OAuth2CredentialFlowState.Builder();
+        }
+
+        throw new IllegalStateException("Unsupported connection factory implementation: " + connectionFactory);
+    }
+
+    static Class<? extends CredentialFlowState> typeForName(final String name) {
```

Comment:
I think I need to give Jackson a concrete type, but I did fiddle with the annotations so it could be leftover from when id did not work for me. I think I need a unit test for that.